File: html5/RadXml/RadXmlLevel.js
/*
	MIT-LICENSE
	Copyright (c) 2013 Higher Edge Software, LLC

	Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
	and associated documentation files (the "Software"), to deal in the Software without restriction, 
	including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
	and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
	subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all copies or substantial 
	portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
	LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION 
	WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

function Level ()
{
	this._name = "";
	this._ambientColour = new Colour (1, 1, 1, 1);
	this._shadowColour = new Colour (0, 0, 0, 1);

	this._aryAnimations = new Array ();
	this._aryMusic = new Array ();
	this._aryOverlays = new Array ();
	this._aryOverlayElements = new Array ();

	this._hshMovables = new Array ();

	this._addEvent (new Event ("onload", ""));

	this.setName = function (newName)
	{
		this._name = newName;
	}

	this.getName = function ()
	{
		return (this._name);
	}

	this._parseLevel = function (file, xmlElement)
	{
		RadXML._gameEngine._setupDefaultSceneManager ();

		this._name = RadXML._getAttribute (xmlElement, "name").value;
		this._setJavascriptToEvent ("onload", RadXML._getAttribute (xmlElement, "onload").value);

		var strSkyBox = RadXML._getAttribute (xmlElement, "skybox").value;
		var strAmbientColor = RadXML._getAttribute (xmlElement, "ambientcolor").value;
		var strShadowColor = RadXML._getAttribute (xmlElement, "shadowcolor").value;

		if (this._name == "")
		{
			var iRandName = RadXML._parseInt (Math.random () * 1000000000);
			this.setName (iRandName);
		}

		if (strSkyBox != "")
			RadXML._gameEngine.setSkyBoxMaterial (strSkyBox);

		if (strAmbientColor != "")
		{
			this._ambientColour = RadXML._parseColourValue (strAmbientColor);
			RadXML._gameEngine.setAmbientLight (this._ambientColour);
		}

		if (strShadowColor != "")
		{
			this._shadowColour = RadXML._parseColourValue (strShadowColor);
			RadXML._gameEngine.setShadowColour (this._ambientColour);
		}

		if (xmlElement.childNodes.length > 0)
			this._parseXMLMovables (file, xmlElement.childNodes[0]);

		this._executeJavascriptFromEvent ("onload");
	}

	this._parseXMLMovables = function (file, element, parent)
	{
		for (var iIdx = 0; iIdx < 9; iIdx++)
		{
			var strType = "";
			var iType = (iIdx + 1);

			switch (iIdx)
			{
				case 0:
					strType = "entity";
					break;
				case 1:
					strType = "camera";
					break;
				case 2:
					strType = "particlesystem";
					break;
				case 3:
					strType = "light";
					break;
				case 4:
					strType = "animationtrack";
					break;
				case 5:
					strType = "audiolistener";
					break;
				case 6:
					strType = "sound";
					break;
				case 7:
					strType = "overlay";
					break;
				case 8:
					strType = "overlayelement";
					break;
			}

			var aryTags = RadXML._getElementsByTagName (element, strType, true);

			for (var iJdx = 0; iJdx < aryTags.length; iJdx++)
			{
				var xmlElement = aryTags[iJdx];

				this._parseXMLMovable (file, iType, xmlElement, parent);
			}
		}
	}

	this._parseXMLMovable = function (file, type, element, parent)
	{
		var objObject = this._createMovable (file, type, "", parent, element);

		if (element.childNodes.length > 0)
			this._parseXMLMovables (file, element.childNodes[0], objObject);
	}

	this._createMovable = function (file, type, name, parent, element, createNow)
	{
		if (createNow == null)
			createNow = true;

		var objObject = null;

		if (type == GameObjectTypes.GOT_ENTITY)
		{
			var eEntity = new Entity (name);

			if (element != null)
				eEntity._parseXML (file, element, parent);

			if (createNow == true)
				eEntity._create (parent);

			objObject = eEntity;
		}

		if (type == GameObjectTypes.GOT_CAMERA)
		{
			var cCamera = new Camera (name);

			if (element != null)
				cCamera._parseXML (file, element, parent);

			if (createNow == true)
				cCamera._create (0, parent);

			objObject = cCamera;
		}

		if (type == GameObjectTypes.GOT_PARTICLE_SYSTEM)
		{
			var psSystem = new ParticleSystem (name);

			if (element != null)
				psSystem._parseXML (file, element, parent);

			if (createNow == true)
				psSystem._create (parent);

			objObject = psSystem;
		}

		if (type == GameObjectTypes.GOT_LIGHT)
		{
			var lLight = new Light (name);

			if (element != null)
				lLight._parseXML (file, element, parent);

			if (createNow == true)
				lLight._create (parent);

			objObject = lLight;
		}

		if (type == GameObjectTypes.GOT_ANIMATION_TRACK)
		{
			var atoObject = new AnimationTrackObject (name);

			if (element != null)
				atoObject._parseXML (file, element, parent);

			if (createNow == true)
				atoObject._create (parent);

			objObject = atoObject;
			this._aryAnimations.push (atoObject);
		}

		if (type == GameObjectTypes.GOT_AUDIO_LISTENER)
		{
			var alListener = new AudioListener (name);

			if (element != null)
				alListener._parseXML (file, element, parent);

			if (createNow == true)
				alListener._create (parent);

			objObject = alListener;
		}

		if (type == GameObjectTypes.GOT_SOUND)
		{
			var sSound = new Sound (name);

			if (element != null)
				sSound._parseXML (file, element, parent);

			if (createNow == true)
				sSound._create (parent);

			objObject = sSound;
		}

		if (type == GameObjectTypes.GOT_OVERLAY)
		{
			var oOverlay = new Overlay (name);

			if (element != null)
				oOverlay._parseXML (file, element, parent);

			if (createNow == true)
				oOverlay._create (parent);

			this._aryOverlays.push (oOverlay);
			objObject = oOverlay;
		}

		if (type == GameObjectTypes.GOT_OVERLAY_ELEMENT)
		{
			var oOverlayElement = new OverlayElement (name);

			if (element != null)
				oOverlayElement._parseXML (file, element, parent);

			if (createNow == true)
				oOverlayElement._create (parent);

			this._aryOverlayElements.push (oOverlayElement);
			objObject = oOverlayElement;
		}

		if (objObject != null)
			this._addMovable (objObject);

		return (objObject);
	}

	this._addMovable = function (newMovable)
	{
		this._hshMovables[newMovable.getName ()] = newMovable;
		RadXML.getSingletonPtr ()._addMovableJSObject (newMovable);
	}
}

Level.prototype = new Events ();

File: source/RadXml/RadXmlLevel.cpp
#include "RadXmlLevel.h"
#include "RadXmlXmlFile.h"

#include "RadXml.h"

#include "RadXmlMovable.h"
#include "RadXmlEntity.h"
#include "RadXmlCamera.h"
#include "RadXmlLight.h"
#include "RadXmlParticleSystem.h"
#include "RadXmlAudioListener.h"
#include "RadXmlSound.h"
#include "RadXmlAnimationTrack.h"
#include "RadXmlOverlay.h"
#include "RadXmlOverlayElement.h"

#include "xml/xml.h"

#include "common/mai_radxml_system.h"
#include "common/com_radxml_strconv.h"

#include "languages/language.h"

#ifdef defAPI_OGRE3D
	#ifdef defAPI_OGRE3D_NOT_IN_DIRECTORY
		#include <OGRE/OgreException.h>
		#include <OGRE/OgreColourValue.h>
		#include <OGRE/OgreSceneManager.h>
	#else
		#include <OgreException.h>
		#include <OgreColourValue.h>
		#include <OgreSceneManager.h>
	#endif
#endif

namespace RadXml
{
	Level::Level ()
	{
		name = "";
		hashName = 0;

		addEvent (new Event ("onload", ""));
	}

	Level::~Level ()
	{
	}

	void Level::setName (std::string newName)
	{
		name = newName;
		hashName = RadXmlCommon::s13StringToHash (newName);
	}

	void Level::addMusic (Sound *music)
	{
		aryMusic.push_back (music);
	}

	void Level::parseLevel (XmlFile *file, XMLElement *element)
	{
		Ogre::SceneManager *mSceneMgr = RadXML::getSingletonPtr ()->getSceneManager ();

		name = element->getAttributeValue ("name");
		setJavascriptToEvent ("onload", element->getAttributeValue ("onload"));

		std::string strSkyBox = element->getAttributeValue ("skybox");
		std::string strAmbientColor = element->getAttributeValue ("ambientcolor");
		std::string strShadowColor = element->getAttributeValue ("shadowcolor");

		if (name == "")
		{
			int iRandName = rand ();
			setName (RadXmlCommon::s13IntToString (iRandName));
		}

		if (strSkyBox != "")
			mSceneMgr->setSkyBox (true, strSkyBox);

		if (strAmbientColor != "")
		{
			Ogre::ColourValue cvColor = RadXML::parseColourValue (strAmbientColor);
			mSceneMgr->setAmbientLight (cvColor);
		}

		if (strShadowColor != "")
		{
			Ogre::ColourValue cvColor = RadXML::parseColourValue (strAmbientColor);
			mSceneMgr->setShadowColour (cvColor);
		}

		parseXMLMovables (file, element);
	}

	void Level::parseXMLMovables (XmlFile *file, XMLElement *element, Movable *parent)
	{
		for (int iIdx = 0; iIdx < 9; iIdx++)
		{
			std::string strType = "";
			int iType = (iIdx + 1);

			switch (iIdx)
			{
				case 0:
					strType = "entity";
					break;
				case 1:
					strType = "camera";
					break;
				case 2:
					strType = "particlesystem";
					break;
				case 3:
					strType = "light";
					break;
				case 4:
					strType = "animationtrack";
					break;
				case 5:
					strType = "audiolistener";
					break;
				case 6:
					strType = "sound";
					break;
				case 7:
					strType = "overlay";
					break;
				case 8:
					strType = "overlayelement";
					break;
			}

			std::vector<XMLElement *> aryTags = element->findAllTags (strType, 0, false);

			for (unsigned int iJdx = 0; iJdx < aryTags.size (); iJdx++)
			{
				try
				{
					parseXMLMovable (file, iType, aryTags.at (iJdx), parent);
				}
				catch (Ogre::Exception &ex)
				{
					RadXmlCommon::s13LogMessage ((std::string)MSG15 + ex.getFullDescription () + 
											(std::string)MSG16 + Ogre::StringConverter::toString (ex.getLine ()) + 
											(std::string)MSG17 + ex.getSource ());
				}
			}
		}
	}

	void Level::parseXMLMovable (XmlFile *file, int type, XMLElement *element, Movable *parent)
	{
		Movable *objObject = createMovable (file, type, "", parent, element);

		if (element->content != "")
			parseXMLMovables (file, element, objObject);
	}

	Movable *Level::createMovable (XmlFile *file, int type, std::string name, Movable *parent, XMLElement *element, bool createNow)
	{
		Movable *objObject = 0;

		if (type == GOT_ENTITY)
		{
			Entity *eEntity = new Entity (name);

			if (element != 0)
				eEntity->parseXML (file, element, parent);

			if (createNow == true)
				eEntity->create (parent);

			objObject = eEntity;
		}

		if (type == GOT_CAMERA)
		{
			Camera *cCamera = new Camera (name);

			if (element != 0)
				cCamera->parseXML (file, element, parent);

			if (createNow == true)
				cCamera->create (0, parent);

			objObject = cCamera;
		}

		if (type == GOT_PARTICLE_SYSTEM)
		{
			ParticleSystem *psSystem = new ParticleSystem (name);

			if (element != 0)
				psSystem->parseXML (file, element, parent);

			if (createNow == true)
				psSystem->create (parent);

			objObject = psSystem;
		}

		if (type == GOT_LIGHT)
		{
			Light *lLight = new Light (name);

			if (element != 0)
				lLight->parseXML (file, element, parent);

			if (createNow == true)
				lLight->create (parent);

			objObject = lLight;
		}

		if (type == GOT_ANIMATION_TRACK)
		{
			AnimationTrackObject *atoObject = new AnimationTrackObject (name);

			if (element != 0)
				atoObject->parseXML (file, element, parent);

			if (createNow == true)
				atoObject->create (parent);

			objObject = atoObject;
			aryAnimations.push_back (atoObject);
		}

		if (type == GOT_AUDIO_LISTENER)
		{
			AudioListener *alListener = new AudioListener (name);

			if (element != 0)
				alListener->parseXML (file, element, parent);

			if (createNow == true)
				alListener->create (parent);

			objObject = alListener;
		}

		if (type == GOT_SOUND)
		{
			Sound *sSound = new Sound (name);

			if (element != 0)
				sSound->parseXML (file, element, parent);

			if (createNow == true)
				sSound->create (parent);

			objObject = sSound;
		}

		if (type == GOT_OVERLAY)
		{
			Overlay *oOverlay = new Overlay (name);

			if (element != 0)
				oOverlay->parseXML (file, element, parent);

			if (createNow == true)
				oOverlay->create (parent);

			aryOverlays.push_back (oOverlay);
			objObject = oOverlay;
		}

		if (type == GOT_OVERLAY_ELEMENT)
		{
			OverlayElement *oOverlayElement = new OverlayElement (name);

			if (element != 0)
				oOverlayElement->parseXML (file, element, parent);

			if (createNow == true)
				oOverlayElement->create (parent);

			aryOverlayElements.push_back (oOverlayElement);
			objObject = oOverlayElement;
		}

		addMovable (objObject);

		return (objObject);
	}

	void Level::addMovable (Movable *newMovable)
	{
		hshMovables.insert (std::pair<unsigned long, Movable *> (newMovable->getHash (), newMovable));
		RadXML::getSingletonPtr ()->getJavascriptEngine ()->addMovableJSObject (newMovable);
	}

	Movable *Level::getMovable (std::string findName, bool skipErrorMessage)
	{
		HASHMAP<unsigned long, Movable *>::iterator itBegin = hshMovables.find (RadXmlCommon::s13StringToHash (findName));
		HASHMAP<unsigned long, Movable *>::iterator itEnd = hshMovables.end ();

		if (itBegin != itEnd)
			return ((*itBegin).second);
		else
		{
			if (skipErrorMessage == false)
				RadXmlCommon::s13LogMessage (MSG29 + findName);
		}

		return (0);
	}

	Movable *Level::getMovable (unsigned long hash, bool skipErrorMessage)
	{
		HASHMAP<unsigned long, Movable *>::iterator itBegin = hshMovables.find (hash);
		HASHMAP<unsigned long, Movable *>::iterator itEnd = hshMovables.end ();

		if (itBegin != itEnd)
			return ((*itBegin).second);
		else
		{
			if (skipErrorMessage == false)
				RadXmlCommon::s13LogMessage (MSG30);
		}

		return (0);
	}

	void Level::removeMovable (std::string findName)
	{
		HASHMAP<unsigned long, Movable *>::iterator itBegin = hshMovables.find (RadXmlCommon::s13StringToHash (findName));
		HASHMAP<unsigned long, Movable *>::iterator itEnd = hshMovables.end ();

		if (itBegin != itEnd)
		{
			RadXML::getSingletonPtr ()->getJavascriptEngine ()->removeMovableJSObject ((*itBegin).second);
			hshMovables.erase (itBegin);
		}
	}

	void Level::removeMovable (unsigned long hash)
	{
		HASHMAP<unsigned long, Movable *>::iterator itBegin = hshMovables.find (hash);
		HASHMAP<unsigned long, Movable *>::iterator itEnd = hshMovables.end ();

		if (itBegin != itEnd)
		{
			RadXML::getSingletonPtr ()->getJavascriptEngine ()->removeMovableJSObject ((*itBegin).second);
			hshMovables.erase (itBegin);
		}
	}
}

